From bebace16580f0a7bee31a7d4d9314a951d8c169a Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:06:30 -0400
Subject: [usb] storage: Modify and export adjust_quirks so that it can be used by uas

Message-id: <1410887463-20674-319-git-send-email-dzickus@redhat.com>
Patchwork-id: 95427
O-Subject: [RHEL7 PATCH 318/591] usb-storage: Modify and export adjust_quirks so that it can be used by uas
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit d24d481b7d369b08cce734bf80be374eed5a6e58
Author: Hans de Goede <hdegoede@redhat.com>
Date:   Sat Nov 16 12:09:39 2013 +0100

    usb-storage: Modify and export adjust_quirks so that it can be used by uas

    Signed-off-by: Hans de Goede <hdegoede@redhat.com>
    Signed-off-by: Sarah Sharp <sarah.a.sharp@linux.intel.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/storage/usb.c b/drivers/usb/storage/usb.c
index 388f567..f1c9626 100644
--- a/drivers/usb/storage/usb.c
+++ b/drivers/usb/storage/usb.c
@@ -463,14 +463,14 @@ static int associate_dev(struct us_data *us, struct usb_interface *intf)
 #define TOLOWER(x) ((x) | 0x20)
 
 /* Adjust device flags based on the "quirks=" module parameter */
-static void adjust_quirks(struct us_data *us)
+void usb_stor_adjust_quirks(struct usb_device *udev, unsigned long *fflags)
 {
  char *p;
- u16 vid = le16_to_cpu(us->pusb_dev->descriptor.idVendor);
- u16 pid = le16_to_cpu(us->pusb_dev->descriptor.idProduct);
+ u16 vid = le16_to_cpu(udev->descriptor.idVendor);
+ u16 pid = le16_to_cpu(udev->descriptor.idProduct);
  unsigned f = 0;
  unsigned int mask = (US_FL_SANE_SENSE | US_FL_BAD_SENSE |
-   US_FL_FIX_CAPACITY |
+   US_FL_FIX_CAPACITY | US_FL_IGNORE_UAS |
    US_FL_CAPACITY_HEURISTICS | US_FL_IGNORE_DEVICE |
    US_FL_NOT_LOCKABLE | US_FL_MAX_SECTORS_64 |
    US_FL_CAPACITY_OK | US_FL_IGNORE_RESIDUE |
@@ -541,14 +541,18 @@ static void adjust_quirks(struct us_data *us)
   case 's':
    f |= US_FL_SINGLE_LUN;
    break;
+  case 'u':
+   f |= US_FL_IGNORE_UAS;
+   break;
   case 'w':
    f |= US_FL_NO_WP_DETECT;
    break;
   /* Ignore unrecognized flag characters */
   }
  }
- us->fflags = (us->fflags & ~mask) | f;
+ *fflags = (*fflags & ~mask) | f;
 }
+EXPORT_SYMBOL_GPL(usb_stor_adjust_quirks);
 
 /* Get the unusual_devs entries and the string descriptors */
 static int get_device_info(struct us_data *us, const struct usb_device_id *id,
@@ -568,7 +572,7 @@ static int get_device_info(struct us_data *us, const struct usb_device_id *id,
    idesc->bInterfaceProtocol :
    unusual_dev->useTransport;
  us->fflags = id->driver_info;
- adjust_quirks(us);
+ usb_stor_adjust_quirks(us->pusb_dev, &us->fflags);
 
  if (us->fflags & US_FL_IGNORE_DEVICE) {
   dev_info(pdev, "device ignored\n");
diff --git a/drivers/usb/storage/usb.h b/drivers/usb/storage/usb.h
index 75f70f0..307e339 100644
--- a/drivers/usb/storage/usb.h
+++ b/drivers/usb/storage/usb.h
@@ -201,4 +201,7 @@ extern int usb_stor_probe1(struct us_data **pus,
 extern int usb_stor_probe2(struct us_data *us);
 extern void usb_stor_disconnect(struct usb_interface *intf);
 
+extern void usb_stor_adjust_quirks(struct usb_device *dev,
+  unsigned long *fflags);
+
 #endif
-- 
1.7.1